library/core/src/iter/traits/unchecked_iterator.rs

```diff
@@ -0,0 +1,36 @@
+use crate::iter::TrustedLen;
+
+/// [`TrustedLen`] cannot have methods, so this allows augmenting it.
+///
+/// It currently requires `TrustedLen` because it's unclear whether it's
+/// reasonably possible to depend on the `size_hint` of anything else.
+pub(crate) trait UncheckedIterator: TrustedLen {
```

Comment:
This seems like another take on TrustedRandomAccess, except that the source can drop its elements. I tried something similar in #93243 except by augmenting TRA directly, but it lead to compile time regressions due to the added traits and methods.
Ideally we would only have to maintain one unsafe iteration trait. But it is a lot simpler and sometimes the single-induction-variable that TRA provides isn't necessary, so maybe it's not so bad.

Reply:
Hmm, what's the best resource to read up on `TrustedRandomAccessNoCoerce`?  TBH I've always ignored it 😬  Is <https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/std/iter/trait.TrustedRandomAccess.html#safety> the go-to place?

I could try changing `try_from_trusted_iterator` to essentially `try_from_fn(Iterator::__iterator_get_unchecked)`, rather needing `next_unchecked` at all...

(But `next_unchecked` has a *much* simpler safety condition than `TRANC`.)

Reply:
> s https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/std/iter/trait.TrustedRandomAccess.html#safety the go-to place?

Yes. But nothing actionable for this PR. The trait as it is currently isn't fully applicable for `array::IntoIter` anyway because it is an owning iterator which needs to drop items when it's not exhausted. So it can only be implemented for `T: Copy` (as we do for `vec::IntoIter`) as approximation for `needs_drop() == false`. To generalize it further would need some changes, as tried in the mentioned PR.

I was just pointing out that we now have two ways for unchecked access. TrustedLen is slightly more general since it can also apply to `Chain` or `Flatten` because it supports len > usize::MAX. TRA on the other hand is a bit more powerful when it applies because it allows external iteration with a single induction variable (this really helps `Zip`) and doesn't need to update internal fields on which the next loop iteration would depend.

> (But next_unchecked has a much simpler safety condition than TRANC.)

Agreed, there's some simplicity to it. I have tried using `next().unwrap_unchecked()` before in conjunction with `TrustedLen` while it brought some improvements it couldn't match TRA.
I hope that maybe `UncheckedIterator` could get us closer (if we turn it into a specialization trait, which I thought it was 😅), perhaps even to parity. If we could do away with the TRA machinery that would be awesome because it's chernobyl-levels of unsafe.

But it doesn't cover all the same cases then we'd be left with two unsafe iteration traits, at least one which we might never be able to stabilize and that seems kind of ugly.

Reply:
I'd be happy to play with it in a future PR :)

What's the best way to make it a specialization trait?  Make the method `__iterator_next_unchecked` instead, like `__iterator_get_unchecked`, so it's on `Iterator` instead and just bounded by the specialization trait `TrustedLen`?  (TBH I'm not sure *why* the unsafe specialization trait keeps things from having methods, but presumably there's a good reason.)

Reply:
Specialization traits can have methods (`TrustedRandomAccessNoCoerce` has `size()`). `__iterator_next_unchecked` was moved to the iterator trait during the specialization -> min_specialization refactor for reasons I have forgotten.

Slapping `#[rustc_specialization_trait]` on it might be enough.

Reply:
Oh, I see -- `#[rustc_specialization_trait]` can have methods, but `TrustedLen` is `#[rustc_unsafe_specialization_marker]`, which can't.  That's the distinction I'd missed.  Thanks!